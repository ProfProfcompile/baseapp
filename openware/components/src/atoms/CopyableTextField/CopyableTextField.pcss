.cr-copyable-text-field {
  border: 1px solid var(--base-3);
  background: var(--background-dark);
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: var(--gap);

  &__button {
    cursor: pointer;
    align-self: right;
    color: var(--base-3);
    background: var(--background-dark);
  }

  &__input {
    flex: 1;
    align-self: left;
    color: var(--base-3);
    background: var(--background-dark);
  }

  input {
    width: 100%;
    border-radius: 0;
    border-style: none;
    color: var(--base-3);
    background: var(--background-dark);
  }
}
