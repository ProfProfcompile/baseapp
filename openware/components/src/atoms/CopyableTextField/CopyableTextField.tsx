import '@openware/cryptofont';

import classnames from 'classnames';
import * as React from 'react';

interface CopyableTextFieldProps {
    /**
     * Text value that will be copied to the clipboard
     */
    value: string;
    /**
     * Additional class name for styling. By default element receives `cr-button` class
     * @default empty
     */
    className?: string;
    /**
     * String value that makes copy field be unique
     */
    fieldId: string;
}

type CopyTypes = HTMLInputElement | null;

const copy = (id: string) => {
    const copyText: CopyTypes = document.querySelector(`#${id}`);

    if (copyText) {
        copyText.select();

        document.execCommand('copy');
        window.getSelection().removeAllRanges();
    }
};

/**
 * Text field component with ability to copy inner text.
 */
class CopyableTextField extends React.Component<CopyableTextFieldProps> {
    public componentDidMount() {
        if (!this.props.fieldId) {
            throw new Error('CopyableTextField must contain `fieldId` prop');
        }
    }

    public render() {
        const { value, className } = this.props;
        const { fieldId } = this.props;
        const doCopy = () => copy(fieldId);
        const cx = classnames('cr-copyable-text-field', className);
        return (
            <div className={cx}>
                <div className="cr-copyable-text-field__input">
                    <input
                        id={String(fieldId)}
                        readOnly={true}
                        type="text"
                        value={value}
                    />
                </div>
                <div className="cr-copyable-text-field__button" onClick={doCopy}>
                    Copy
                </div>
            </div>
        );
    }
}

export {
    CopyableTextField,
    CopyableTextFieldProps,
};
