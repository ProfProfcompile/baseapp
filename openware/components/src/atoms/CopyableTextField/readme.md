Example of usage:

```js
  <CopyableTextField fieldId="id1" value="1F1tAaz5x1HUXrCNLbtMDqcw6o5GNn4dfE" />
```
