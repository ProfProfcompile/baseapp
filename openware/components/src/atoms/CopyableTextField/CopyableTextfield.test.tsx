import { shallow } from 'enzyme';
import * as React from 'react';
import { SinonStub, stub } from 'sinon';
import * as uuid from 'uuid';
import { CopyableTextField, CopyableTextFieldProps } from './CopyableTextField';

const ID = '0xdeadbeef';
const defaultProps: CopyableTextFieldProps = {
    value: '1F1tAaz5x1HUXrCNLbtMDqcw6o5GNn4dfE',
    fieldId: 'copy_id',
};

const setup = (props: Partial<CopyableTextFieldProps> = {}) =>
    shallow(<CopyableTextField {...{ ...defaultProps, ...props }} />);

describe('CopyableTextField', () => {
    let uuidMock: SinonStub;

    beforeAll(() => {
        uuidMock = stub(uuid, 'v4').returns(ID);
    });

    afterAll(() => {
        uuidMock.reset();
    });

    it('should render', () => {
        const wrapper = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        const wrapper = setup();
        expect(wrapper.hasClass('cr-copyable-text-field')).toBeTruthy();
    });

    it('should render 1 input tag', () => {
        const wrapper = setup();
        const input = wrapper.find('input');
        expect(input.length).toBe(1);
    });

    it('should render 3 divs', () => {
        const wrapper = setup();
        const divs = wrapper.find('div');
        // tslint:disable-next-line:no-magic-numbers
        expect(divs.length).toBe(3);
    });

    it('should pass along supplied className', () => {
        const className = 'some-class';
        const wrapper = setup({ className });
        expect(wrapper.hasClass(className)).toBeTruthy();
    });

    it('should render correct className for input field', () => {
        const wrapper = setup();
        expect(wrapper.find('div').get(1).props.className)
            .toContain('cr-copyable-text-field__input');
    });

    it('should render correct className for clickable block', () => {
        const wrapper = setup();
        const { className } = wrapper.find('div').last().props();
        expect(className).toContain('cr-copyable-text-field__button');
    });
});
