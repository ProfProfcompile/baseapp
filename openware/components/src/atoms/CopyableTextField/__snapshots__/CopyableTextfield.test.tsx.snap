// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyableTextField should render 1`] = `
<div
  className="cr-copyable-text-field"
>
  <div
    className="cr-copyable-text-field__input"
  >
    <input
      id="copy_id"
      readOnly={true}
      type="text"
      value="1F1tAaz5x1HUXrCNLbtMDqcw6o5GNn4dfE"
    />
  </div>
  <div
    className="cr-copyable-text-field__button"
    onClick={[Function]}
  >
    Copy
  </div>
</div>
`;
